### ISSUE TITLE
Consistent translation of queues?

### ISSUE THREAD

**Issue by danielmarbach** [Particular/NServiceBus#4948]
It seems we are applying different standards when it comes to queue translation. For example the [shared queue](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/Receiving.cs#L24) and the [instance specific queue](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/Receiving.cs#L29) are translated by calling `ToTransportAddress`. These queues are also passed to the `QueueBindings` which are passed to the installer. The [error queue](https://github.com/Particular/NServiceBus/blob/197bea131200dea086b322d33748622e1656e548/src/NServiceBus.Core/Recoverability/Recoverability.cs#L38), the [forwarding queue](https://github.com/Particular/NServiceBus/blob/d462080b0efa6609f0f92d4e192d760142ac47c2/src/NServiceBus.Core/Forwarding/ForwardReceivedMessages.cs#L25) and the [audit queue](https://github.com/Particular/NServiceBus/blob/f9d697cd65e7e9b774a7022bf59428094318526e/src/NServiceBus.Core/Audit/Audit.cs#L31) are not translated but passed into the QueueBindings which are used by the installers. The implications are:

- Input queues are translated
- Output queues are not translated
- `QueueBindings` contains a mixture of translated and not translated queues

This means a transport needs to apply queue translation in the receiver, installer and dispatcher. Should we consistently translate the queues as a service provided by the core? Or should we consistently NOT translate the queues to avoid the in between state?

**Comment by SzymonPobiega** [comment]
The reason for the lack of consistency is the fact that some things like input queue, error queue or the destination set via `SendOptions` are transport-level addresses provided by the users while others (input queues) are machine-generated based on the endpoint name via `ToTransportAddress`. 

There are multiple valid forms of transport addresses provided by the users e.g. for SQL Server transport valid forms are `error`, `error@[dbo]` and `error@[dbo]@MyCatalog`. Provided that initial catalog in the endpoint's connection string is `MyCatalog` these addresses are equivalent and can be used interchangably. The dispatcher will substitute missing parts with defaults/configuration.

There is, however, a distinct *canonical form* for each transport. For SQL it is `error@[dbo]@MyCatalog`. It can be obtain for each address by calling `MakeCanonicalForm`.

After discussion this on Slack with @danielmarbach (full log available on request) my proposal is following
 * Change the return type of `MakeCanonicalForm` and `ToTransportAddress` to object. Let these methods return a transport-specific address object
 * Require passing of this transport-specific address in the dispatcher, the message pump and the queue creator (instead of a string as in current version)

**Comment by bording** [comment]
Does anything need to be changed here at all? The idea is that everything being passed into `QueueBindings` are in fact fully translated addresses. Some of those are logical endpoint names that were turned into addresses via `ToTransportAddress`, but others are provided directly by the user, as @SzymonPobiega has mentioned. Users are supposed to provide valid addresses, so changing those further or passing them through `ToTransportAddress` doesn't make sense.

**Comment by SzymonPobiega** [comment]
@bording the idea was to pass them to `MakeCanonicalForm` for validation. This way we would also centralize the place where we parse (possibly multiple forms of) addresses and only deal with valid, parsed ones in dispatcher, pump and queue creator.

That said, I don't think we *need* to do this. I think we should consider it next time we open the transport seam for changes.

**Comment by danielmarbach** [comment]
@bording I don't think we necessarily need to change anything. I brought this up to have a discussion. For example I did not remember that we want to force people into giving valid queue names for the error queue and audit queue. For example if queue name shortening is applied in a transport the input queue is automatically shorted but if the error queue is too long or does not follow the convention the transport will throw. Not saying that is necessarily bad. Just something I had to remind myself and I was thinking whether we could improve the story here a bit.
